### PR TITLE
Form should support on-open for nested objects

### PR DESCRIPTION
**Pull request by eoaksnes**
## What does this pull request change?

* Be able to call on-open for 
  * nested objects 
  * list of objects
* Be able to create optional nested objects that do not have any value
* Be able to select or create non-contained nested objects
* Be able to remove non-contained objects 
* Be able to remove contained optional objects

Also included:
* Be able to sort ordering of attributes 
* Be able to specify date-time format for string attribute

image

## Why is this pull request needed?

* To support tab container

### CODE REVIEW

File: web/packages/plugins/shared/common/src/components/TreeView.tsx [hunk 1]
```diff
@@ -130,17 +131,16 @@ export const TreeView = (props: {
   }, [])
 
   const _onClick = (node: TreeNode) => {
-    if (!node.expanded) {
+    if (![BlueprintEnum.PACKAGE, 'dataSource'].includes(node.type)) {
+      onSelect(node)
+    } else if (!node.expanded) {
```
Comment (soofstad) [inline web/packages/plugins/shared/common/src/components/TreeView.tsx:136]: This seem to have broken collapsable DataSources in the treeView

Comment (soofstad) [reply web/packages/plugins/shared/common/src/components/TreeView.tsx:136]: What was the problem this was trying to fix?

File: web/packages/plugins/shared/common/src/components/UiPluginSelector.tsx [hunk 2]
```diff
@@ -224,6 +225,7 @@ export function UIPluginSelector(props: {
           onOpen={onOpen}
           categories={categories}
           config={config}
+          type={entity.type}
```
Comment (soofstad) [inline web/packages/plugins/shared/common/src/components/UiPluginSelector.tsx:228]: Type should be removed from the UiPlugin interface.
As a requirement is that it passes an entity which MUST have a type

File: web/packages/plugins/shared/common/src/hooks/useBlueprint.tsx [hunk 3]
```diff
@@ -6,6 +6,7 @@ export const useBlueprint = (typeRef: string) => {
   const [blueprint, setBlueprint] = useState<any | null>(null)
   const [isLoading, setLoading] = useState<boolean>(true)
   const [error, setError] = useState<Error | null>(null)
+  // @ts-ignore
```
Comment (soofstad) [inline web/packages/plugins/shared/common/src/hooks/useBlueprint.tsx:9]: These ignores should not be needed

File: web/packages/plugins/ui/tabs/src/TabsContext.tsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import React, { createContext, useContext } from 'react'
+
+// @ts-ignore
+export const TabsContext = createContext<any>({
+  fields: {},
+  widgets: {},
+})
```
Comment (soofstad) [inline web/packages/plugins/ui/tabs/src/TabsContext.tsx:7]: Did we not move away from having Tabs as a Provider?
Also why do Tabs care or need to know about fields and widgets?

**Review by soofstad — CHANGES_REQUESTED** [review]

File: api/src/home/analysis-platform/data/WorkflowDS/Blueprints/Task.json [hunk 5]
```diff
@@ -43,6 +43,22 @@
...
+        "type": "DMT-Internal/DMT/ui-plugins/form/FormInput",
+        "attributes": [
+          {
+            "type": "DMT-Internal/DMT/ui-plugins/form/fields/StringField",
+            "name": "outputType",
+            "widget": "BlueprintPickerWidget"
+          }
+        ]
+      }
+    },
```
Comment (soofstad) [inline api/src/home/analysis-platform/data/WorkflowDS/Blueprints/Task.json:61]: Should remove "Form" from the Blueprints where we have custom plugins. And set "Form" as the default edit plugin for everything instead. Which we override in cases like Task.json

File: web/packages/plugins/shared/common/src/components/TreeView.tsx (hunk 1 above)
Comment (eoaksnes) [reply web/packages/plugins/shared/common/src/components/TreeView.tsx]: I got some memory leak issue, but I did not manage to reproduce it, so remove the changes again.

File: web/packages/plugins/shared/common/src/components/UiPluginSelector.tsx (hunk 2 above)
Comment (eoaksnes) [reply web/packages/plugins/shared/common/src/components/UiPluginSelector.tsx]: okey, removed

File: web/packages/plugins/ui/tabs/src/TabsContext.tsx (hunk 4 above)
Comment (eoaksnes) [reply web/packages/plugins/ui/tabs/src/TabsContext.tsx]: I think we can keep the provider for tabs, so we don't need to pass down the onOpen through all the nested children. I will remove the fields and widgets, copy paste error.

File: api/src/home/analysis-platform/data/WorkflowDS/Blueprints/Task.json (hunk 5 above)
Comment (eoaksnes) [reply api/src/home/analysis-platform/data/WorkflowDS/Blueprints/Task.json]: true, will remove

**Review by soofstad — APPROVED** [review]

**Pull request merged**